lib/utils/aws/organizations/matchingaccounts.go

```diff
@@ -0,0 +1,288 @@
+	}
+
+	if len(m.ExcludeOUs) > 1 && slices.Contains(m.ExcludeOUs, "*") {
+		return trace.BadParameter("ExcludeOUs cannot contain '*' along with other OU IDs")
+	}
+
+	return nil
+}
+
+func (m *MatchingAccountsFilter) isIncludeNothing() bool {
+	return len(m.ExcludeOUs) == 1 && m.ExcludeOUs[0] == "*"
+}
+
+func (m *MatchingAccountsFilter) isIncludeAll() bool {
+	return len(m.ExcludeOUs) == 0 && (len(m.IncludeOUs) == 0 || (len(m.IncludeOUs) == 1 && m.IncludeOUs[0] == "*"))
+}
+
+// MatchingAccounts returns the list of account IDs that are part of the organization and match the filter.
+// Every OU in ExcludeOUs is excluded from the results, including its children OUs.
+func MatchingAccounts(ctx context.Context, orgsClient OrganizationsClient, filter MatchingAccountsFilter) ([]string, error) {
+	if err := filter.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if filter.isIncludeNothing() {
+		return []string{}, nil
+	}
+
+	if filter.isIncludeAll() {
+		accountIDs, err := allAccounts(ctx, orgsClient, filter.OrganizationID)
+		return accountIDs, trace.Wrap(err)
+	}
+
+	// General case: build the organization tree, this tree already removes all the excluded OUs and their children OUs.
+	orgTree, err := buildOrgTree(ctx, orgsClient, filter)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Then: iterate over the tree and collect the included accounts.
+	includeEverything := len(filter.IncludeOUs) == 0 || filter.IncludeOUs[0] == "*"
+	return collectIncludedAccounts(orgTree, filter.IncludeOUs, includeEverything), nil
+}
+
+func collectIncludedAccounts(orgItem *awsOrgItem, included []string, isParentIncluded bool) []string {
+	includeCurrentOU := isParentIncluded || slices.Contains(included, orgItem.id)
+
+	var accountIDs []string
+	if includeCurrentOU {
+		accountIDs = append(accountIDs, orgItem.accounts...)
+	}
+
+	for _, orgUnit := range orgItem.organizationalUnits {
+		childAccountIDs := collectIncludedAccounts(orgUnit, included, includeCurrentOU)
+		accountIDs = append(accountIDs, childAccountIDs...)
+	}
+
+	return accountIDs
+}
+
+func allAccounts(ctx context.Context, orgsClient OrganizationsClient, organizationID string) ([]string, error) {
+	var accountIDs []string
+	organizationIDValidated := false
+
+	paginator := organizations.NewListAccountsPaginator(orgsClient, &organizations.ListAccountsInput{})
+
+	for paginator.HasMorePages() {
+		page, err := paginator.NextPage(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		for _, account := range page.Accounts {
+
+			// This check ensures the assumed role belongs to the expected Organization.
+			// Only need to validate once because all accounts will belong to the same Organization.
+			if !organizationIDValidated {
+				if err := accountARNBelongsToOrganizationID(aws.ToString(account.Arn), organizationID); err != nil {
+					return nil, trace.Wrap(err)
+				}
+				organizationIDValidated = true
+			}
```

Comment:
Do we need to make this check? What would it mean for the org to be mismatched?

I'm worried that we are second-guessing AWS here and there might be some strange cases where this validation fails, which in turn breaks the process for entire organization.